src/service/angular-sails.js

```diff
@@ -1,74 +1,168 @@
-                socket[methodName] = function (url, data, cb) {
-                    var deferred = defer();
-                    if (cb === undefined && angular.isFunction(data)) {
-                        cb = data;
-                        data = null;
-                    }
-                    deferred.promise.then(cb);
-                    socket['legacy_' + methodName](url, data, function (result) {
-                        resolveOrReject(deferred, result);
-                    });
-                    return deferred.promise;
-                };
-            },
-            wrapEvent = function (eventName) {
-                socket['legacy_' + eventName] = socket[eventName];
-                socket[eventName] = function (event, cb) {
-                    if (cb !== null && angular.isFunction(cb)) {
-                        socket['legacy_' + eventName](event, function (result) {
-                            angularify(cb, result);
-                        });
-                    }
-                };
-            };
-
-        angular.forEach(httpVerbs, promisify);
-        angular.forEach(eventNames, wrapEvent);
-
-        return socket;
+    this.base = '';
+    this.config = {};
+    this.debug = false;
+
+    // like https://docs.angularjs.org/api/ng/service/$http#interceptors
+    // but with sails.io arguments
+    var interceptorFactories = this.interceptors = [
+      /*function($injectables) {
+         return {
+           request: function(config) {},
+           response: function(response) {},
+           requestError: function(config) {},
+           responseError: function(response) {}
+         };
+       }*/
+    ];
+
+    this.$get = ['$q', '$injector', '$rootScope', '$log', function($q, $injector, $rootScope, $log) {
+      var socket = (io.sails || io).connect(provider.url, provider.config);
+
+      // build interceptor chain
+      var reversedInterceptors = [];
+      angular.forEach(interceptorFactories, function(interceptorFactory) {
+        reversedInterceptors.unshift(
+          angular.isString(interceptorFactory) ?
+            $injector.get(interceptorFactory) : $injector.invoke(interceptorFactory));
+      });
+
+      // Send the request using the socket
+      function serverRequest(config) {
+        var defer = $q.defer();
+        if(provider.debug) $log.info('$sails ' + config.method.toUpperCase() + ' ' + config.url, config.data || '');
+
+        socket['legacy_' + config.method](config.url, config.data, function(result, jwr) {
+          // resolve promise if JSON web response is an object that has a statusCode 2xx
+          jwr.data = result; // backward compat, jwr.body also holds your data
+          jwr.socket = socket;
+          jwr.url = config.url;
+          jwr.method = config.method;
+          jwr.config = config.config;
+          if(!jwr || !angular.isObject(jwr) || !jwr.statusCode || jwr.statusCode < 200 || jwr.statusCode >= 300) {
```

Comment:
Your setting `jwr` before this, why all the checks?  Also, if on the newer version of sails.io we can shortcut this by checking for jwr.error.  See this PR that is pending https://github.com/janpantel/angular-sails/pull/43/files (which I now realize is completely depending on that error coming from `jwr`/new sails.io and would need a fallback in that compatibility section)